Kernel/Devices/AHCIPort.h

```diff
@@ -0,0 +1,187 @@
+    u32 index() const { return m_index; }
+
+private:
+    void setup_command(AHCIPort& port);
+    bool execute_command(AHCIPort& port);
+    void rebase(AHCIPort& port);
+
+    VirtualAddress m_base_addr;
+    u32 m_index;
+    u32 m_ctba_offset;
+
+    struct PRDT_ENTRY {
+        volatile u32 dba;
+        volatile u32 dbau;
+        volatile u32 rsv0;
+
+        volatile u32 dbc : 22;
+        volatile u32 rsv1 : 9;
+        volatile u32 i : 1;
+    };
+
+    struct FIS_REG_H2D {
+        volatile u8 fis_type;
+
+        volatile u8 pmport : 4;
+        volatile u8 rsv0 : 3;
+        volatile u8 mode : 1;
+
+        volatile u8 command;
+        volatile u8 featurel;
+
+        volatile u8 lba0;
+        volatile u8 lba1;
+        volatile u8 lba2;
+        volatile u8 device;
+
+        volatile u8 lba3;
+        volatile u8 lba4;
+        volatile u8 lba5;
+        volatile u8 featureh;
+
+        volatile u8 countl;
+        volatile u8 counth;
+        volatile u8 icc;
+        volatile u8 control;
+
+        volatile u8 rsv1[4];
+    };
+
+    struct DW0 {
+        volatile u8 cfl : 5;
+        volatile u8 atapi : 1;
+        volatile u8 write : 1;
+        volatile u8 prefetch : 1;
+
+        volatile u8 reset : 1;
+        volatile u8 bist : 1;
+        volatile u8 clear : 1;
+        volatile u8 reserved : 1;
+        volatile u8 port_mult : 4;
+    };
+    volatile DW0* m_dw0;
+    volatile u16* m_prdtl;
+    volatile u32* m_prdbc;
+    volatile u32* m_ctba;
+    volatile u32* m_ctbau;
+
+    union {
+        u8* m_cfis;
+        FIS_REG_H2D* m_cfis_reg_h2d;
+    };
+    volatile u8* m_acmd;
+    volatile PRDT_ENTRY* m_entries;
+};
+
+class AHCIPort {
+    AK_MAKE_ETERNAL
+public:
+    enum Type {
+        SATA,
+        SATAPI,
+        SEMB,
+        PM,
+        NULL
+    };
+
+    AHCIPort(VirtualAddress base, u32 index);
+
+    void rebase();
+    bool is_busy();
+    bool is_executing(u32 slot);
+    bool has_file_error();
+    void issue_command(u32 slot);
+    void wait_for_irq();
+    AHCISlot& find_free_slot();
+
+    u32 is() { return *m_is; }
+    const String& name() const { return m_name; }
+    u32 size() const { return m_cylinders * m_heads * m_sectors_per_track * 512; }
+
+    u32 index() const { return m_index; }
+    uintptr_t base() const { return m_ahci_base.get(); }
+    uintptr_t base_physical() const { return m_ahci_base_physical.get(); }
+    const Type& type() const { return m_type; }
+
+private:
+    Type check_type();
+    void identify_device();
+    void stop_command_execution();
+    void start_command_execution();
+
+    // Device mata info
+    OwnPtr<Region> m_ahci_base_region;
+    VirtualAddress m_ahci_base;
+    PhysicalAddress m_ahci_base_physical;
+    VirtualAddress m_base_addr;
+    u32 m_index;
+    Type m_type;
+    String m_name;
+    u8 m_cylinders;
+    u8 m_heads;
+    u8 m_sectors_per_track;
+
+    u32 m_clb_offset;
+    u32 m_fb_offset;
+
+    // Data
+    volatile u32* m_clb;
```

Comment:
Why do we need to use volatile pointers? Please map a region in the virtual address space and use that instead with a suitable structure.